Sources/StreamChat/Models/Channel+Sorting.swift

```diff
@@ -0,0 +1,166 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+// last_updated, last_message_at, updated_at, created_at, member_count, unread_count or has_unread
+
+class Example {
+    let channels: [ChatChannel] = []
+
+    func sortWithDynamicMember(with sorting: [Sorting<ChannelListSortingKey>]) -> [ChatChannel] {
+        channels.sorted { lhs, rhs in
+            for sort in sorting {
+                let stringKeyPath = sort.key.rawValue
+                let lhsValue = lhs[dynamicMember: stringKeyPath] ?? ""
+                let rhsValue = rhs[dynamicMember: stringKeyPath] ?? ""
+
+                if sort.isAscending, lhsValue < rhsValue {
+                    return true
+                } else if !sort.isAscending, lhsValue > rhsValue {
+                    return true
+                } else {
+                    continue
+                }
+            }
+
+            return false
+        }
+    }
+
+    struct SortValue<T> {
+        let keyPath: PartialKeyPath<T>
+        let isAscending: Bool
+    }
+
+    func sortWithKeyPath(with sorting: [SortValue<ChatChannel>]) -> [ChatChannel] {
```

Comment:
This solution has the safety of using PartialKeyPath<T> to both list the possible sorting keys, and to get those values from the object in a nice and safe way. But:
- It returns values as `Any`, and then we need to do downcasting
- It does not allow extra data -> I could not think of a way to overcome this, unfortunately. Which kills this approach 🙃

Reply:
Yeah, the only way this could work is to only allow Primitive types as a KeyPath, example:

`custom(key: String, keyPath: KeyPath<ChatChannel, String>)`
`custom(key: String, keyPath: KeyPath<ChatChannel, Int>)`
`custom(key: String, keyPath: KeyPath<ChatChannel, Double>)`
`custom(key: String, keyPath: KeyPath<ChatChannel, Bool>)`

If we convert the `ChannelListSortingKey` to struct, or just add this additional cases, it could potentially work